### PR TITLE
Width-over-depth 4L/768d/12h: bold capacity increase (Issue #18)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Increasing model width from 512d → 768d (keeping 4 layers) will push capacity orthogonally to the failed 6L depth direction, improve τ_y/τ_z representation, and beat the current yi SOTA val_abupt=7.5373%.

**Why width rather than depth:**
- Depth scaling to 6L was tried three times independently (PR #633, PR #645, PR #655) and failed at EP1 gate every time (18.9%, 18.81%, ~19% all >> 15%). The root cause is LR cascade instability: deeper residual stacks require lower per-layer learning rates, but Lion at lr=1e-4 on 6 layers drives the later layers too hard. The direction is closed.
- **Width scaling is orthogonal.** A 4L/768d model keeps the same residual depth (stable Lion dynamics) while expanding the embedding space. Each layer has more representational capacity. The 4L/512d architecture has been stable at Lion lr=1e-4 throughout the campaign — this change preserves that stability.
- The τ_y and τ_z gaps (9.87% and 11.25% val vs 3.65% and 3.63% AB-UPT targets) suggest the model is under-parameterized for the high-frequency surface geometry variations that drive these shear components. More width → more basis functions → better τ_y/τ_z representation.
- You yourself (in PR #645 closing analysis) suggested 4L/768d as the natural follow-up after the depth direction was exhausted.
- This is a bold architecture change per Issue #18 mandate — not incremental tuning.

**Parameter count estimate:**
- 4L/512d/8h (current): ~12.7M params
- 4L/768d/12h (proposed): ~28–30M params (~2.3× capacity increase)
- Heads scale proportionally: 512d/8h = 64d per head → 768d/12h = 64d per head (same head_dim)

## Instructions

This is a cold-start run from scratch (not resumed from a checkpoint) using the full yi base config, with only `--model-hidden-dim` and `--model-heads` changed.

```bash
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --standalone --nproc_per_node=4 train.py \
  --agent norman \
  --wandb-group yi-round37-width-768d \
  --wandb-name "norman/4l-768d-12h-cold-start" \
  --epochs 3 \
  --learnable-pe --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --lr-warmup-epochs 1 \
  --grad-ema-alpha 0.5 \
  --no-compile-model \
  --model-layers 4 --model-hidden-dim 768 --model-heads 12 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --ema-decay 0.999
```

**Key changes vs yi baseline config:**
- `--model-hidden-dim 768` (was 512)
- `--model-heads 12` (was 8; keeps 64d per head, same as 512d/8h)
- `--model-slices 128` (unchanged)
- Everything else identical to yi recommended base config

**EP1 gate (bold architecture):** val_abupt ≤ 15% = continue, 15–25% = marginal/check-in, > 25% = abort immediately. Post an EP1 status comment after epoch 1 completes.

**If EP1 passes (≤15%):** Continue to full budget (3 epochs). If EP1 is strong (≤12%), consider posting for early review.

**Memory note:** 4L/768d with batch_size=4 and 65536 surface+volume points should comfortably fit 4×H100 96GB. If OOM occurs, reduce `--model-slices` to 64 first, then as a last resort try `--batch-size 2`.

**If batch memory is tight:** Try `--batch-size 2 --gradient-accumulation-steps 2` to maintain effective batch size 4.

## Baseline (current yi SOTA — PR #637 fern, val_abupt=7.5373%)

| Metric | val | test |
|---|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **7.5373%** | 8.8533% |
| `surface_pressure_rel_l2_pct` | 4.9322% | 4.6968% |
| `wall_shear_rel_l2_pct` | 8.4774% | 8.4954% |
| `volume_pressure_rel_l2_pct` | 4.4102% | 11.4599% |
| `wall_shear_x_rel_l2_pct` | 7.2272% | 7.3046% |
| `wall_shear_y_rel_l2_pct` | 9.8691% | 9.8648% |
| `wall_shear_z_rel_l2_pct` | 11.2477% | 10.9403% |

W&B run: `vzprvtaw` | Baseline branch: `yi`

**Win criterion: val_abupt < 7.5373%**

## Background context

Three 6L depth-scaling attempts all failed at EP1 (>15%):
- PR #633 (norman clean depth ablation): 18.9% EP1 → aborted
- PR #645 (norman clean 6L with no extras): 18.81% EP1 → aborted  
- PR #655 (senku 6L on full yi stack): ~19% EP1 → aborted

**Width is orthogonal.** It does not add residual depth, does not compound LR cascade instability, and directly increases representational capacity in every layer simultaneously.

## Terminal result marker

When complete, post a terminal result comment:
```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val/abupt_axis_mean_rel_l2_pct","value":<number>},"test_metric":{"name":"test/abupt_axis_mean_rel_l2_pct","value":<number>}}
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT norman: Initial launch hit OOM on rank 0 at MLP forward (4×97GB GPUs / RTX PRO 6000 Blackwell, model 28.45M params). Took ~93.5 GiB before OOM trying to allocate next 768 MiB intermediate. W&B run wp6g3kjz aborted at step 0. Will retry with --model-slices 64 mitigation per PR instructions.

**Comment by morganmcg1** [comment]
ADVISOR: The OOM at rank 0 step 0 with 4L/768d/12h (run `wp6g3kjz`) is expected — 768d×12h is too large for 96GB VRAM at the default batch configuration. Retry with gradient checkpointing via `--model-slices 64` to trade compute for memory.

**Updated launch command:**
```bash
torchrun --standalone --nproc_per_node=4 train.py \
  --num-layers 4 \
  --hidden-dim 768 \
  --num-heads 12 \
  --model-slices 64 \
  --wandb-group yi-round37-depth-scaleup \
  --wandb-run-name yi-r37-4L-768d-12h-sliced
```

If that still OOMs, try `--model-slices 32` (more aggressive slicing). Report the EP1 result once it completes.

**Comment by morganmcg1** [comment]
STUDENT norman: EP1 status check-in (per PR gate spec).

## Run setup
- W&B run ID: `q03gty6i` (group: `yi-round37-width-768d`, name: `norman/4l-768d-12h-cold-start-slices64`)
- Architecture: 4L / 768d / 12h / 64 slices (per advisor mitigation comment)
- Optimizer: Lion lr=1e-4, wd=5e-4, clip=0.5, grad-EMA α=0.5, lr-warmup-epochs=1
- Started: 2026-05-05 03:03 UTC
- 4×RTX PRO 6000 Blackwell (97GB)

## EP1 result
| Metric | Value |
|---|---:|
| `val/abupt_axis_mean_rel_l2_pct` | **15.9627%** |
| `val/surface_pressure_mae` | 0.03110 |
| `val/volume_pressure_mae` | 24.83132 |
| `val/wall_shear_mae` | 0.16470 |
| `train/loss` (final EP1) | 0.32223 |
| EP1 wall time | 9210 s (~153 min) |
| Peak mem | 86.2 GB / 96 GB |

## Gate decision
- PR thresholds: ≤15% continue, 15–25% marginal/check-in, >25% abort.
- 15.9627% is **just over the strict pass band**, sitting in the marginal/check-in band.

## Continuation rationale
I'm letting the run keep going (currently ~3% into EP2) for these reasons:
1. **Comfortably below the abort threshold (>25%)**, and well below the failed 6L EP1 numbers (18.81–19%). The width direction is *not* showing the same instability signature as depth.
2. **2.3× larger model** with the same `--lr-warmup-epochs 1` schedule — first-epoch lag is expected because a bigger model takes longer for representations to settle. The trajectory in EP2 is the more informative signal.
3. **Time budget**: 360 min total, 155 min used. EP2 finishes around 08:09 UTC (~6 min before timeout). EP3 will be cut off, so the most signal we'll get is EP2's `val_abupt`. If EP2 closes the gap meaningfully (<10%), it's a real signal width matters; if EP2 stalls at ~15%, the architecture is plausibly capacity-limited at this LR.

## Pending advisor input
If you'd prefer I **abort now** (e.g. to retask the GPUs), I'll do so on receipt — but with no other PR queued and a clean stability signal at EP1, I'd recommend letting EP2 land. Will post EP2 result the moment it's available, plus the partial EP3 if any.

**Comment by morganmcg1** [comment]
ADVISOR: EP1=15.9627% is below the 25% hard abort threshold and your reasoning is sound — the 4L→768d capacity jump (50% more width, 3× parameters) naturally lags at EP1 before the extra capacity pays off. Let EP2 land.

**EP2 gate: ≤ 10.5% to continue, > 10.5% to close.**

If EP2 passes, post the per-component breakdown (surface_pressure, tau_x/y/z, vol_pressure) and we will assess whether the width is recovering the τ_y/τ_z gap specifically. That is the target metric for this experiment — overall abupt_axis_mean is a coarser signal here.

**Comment by morganmcg1** [comment]
STUDENT norman:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["q03gty6i"],"primary_metric":{"name":"val/abupt_axis_mean_rel_l2_pct","value":10.0258},"test_metric":{"name":"test/abupt_axis_mean_rel_l2_pct","value":11.2020}}

## Results

Run finished. **EP2 best val_abupt = 10.0258%** — passes the 10.5% gate but is +2.49pp worse than the yi SOTA baseline (7.5373%). Hypothesis is **not validated within the 360-min budget**: the τ_y/τ_z gap that motivated the width direction actually *widened*, not closed.

### Run setup
- W&B run: [q03gty6i](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml/runs/q03gty6i) (group `yi-round37-width-768d`, name `norman/4l-768d-12h-cold-start-slices64`)
- Architecture: 4L / hidden=768 / heads=12 / **slices=64** (advisor mitigation for OOM)
- Optimizer: Lion lr=1e-4, wd=5e-4, clip=0.5, grad-EMA α=0.5, lr-warmup-epochs=1
- Hardware: 4×RTX PRO 6000 Blackwell (97 GB)
- Wall-time: 277.3 min / 360 min budget; epoch_time≈109 min/epoch
- **Best epoch: 2 / 3** — EP3 partially trained but did not improve

### Validation metrics (best epoch=2)

| Metric | val | test |
|---|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **10.0258%** | **11.2020%** |
| `surface_pressure_rel_l2_pct` | 6.4796% | 6.2306% |
| `wall_shear_rel_l2_pct` | 11.0443% | 11.0458% |
| `volume_pressure_rel_l2_pct` | 6.6687% | 13.2555% |
| `wall_shear_x_rel_l2_pct` | 9.3270% | 9.4438% |
| `wall_shear_y_rel_l2_pct` | 13.3022% | 13.2822% |
| `wall_shear_z_rel_l2_pct` | 14.3513% | 13.7980% |

### Comparison vs yi SOTA (PR #637, fern, val_abupt=7.5373%)

| Metric | Baseline val | This val | Δ val | Baseline test | This test | Δ test |
|---|---:|---:|---:|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | 7.5373% | 10.0258% | **+2.49** ❌ | 8.8533% | 11.2020% | +2.35 |
| `surface_pressure` | 4.9322% | 6.4796% | +1.55 | 4.6968% | 6.2306% | +1.53 |
| `wall_shear` | 8.4774% | 11.0443% | +2.57 | 8.4954% | 11.0458% | +2.55 |
| `volume_pressure` | 4.4102% | 6.6687% | +2.26 | 11.4599% | 13.2555% | +1.80 |
| `τ_x` | 7.2272% | 9.3270% | +2.10 | 7.3046% | 9.4438% | +2.14 |
| `τ_y` | 9.8691% | 13.3022% | **+3.43** | 9.8648% | 13.2822% | +3.42 |
| `τ_z` | 11.2477% | 14.3513% | **+3.10** | 10.9403% | 13.7980% | +2.86 |

### What happened

**The width-targets-shear hypothesis is not supported by this run.** The τ_y and τ_z gaps were the worst-affected metrics — the *opposite* of what the hypothesis predicted. Within the 360-min budget the wider model never closed the gap to baseline; instead it produced uniformly worse numbers across every component.

A few mechanistic observations:

1. **Convergence was incomplete.** All val slopes remained strongly negative at terminal:
   - `val_primary/abupt_axis_mean_rel_l2_pct`: **−1.54%/1k steps** (still descending)
   - `val_primary/wall_shear_y_rel_l2_pct`: **−1.74%/1k steps**
   - `val_primary/wall_shear_z_rel_l2_pct`: **−2.08%/1k steps**

   The model was clearly still learning. Linear extrapolation of −1.54%/1k×~6k more steps ≈ another −9pp on val_abupt, which is unphysical but shows convergence had not plateaued. With 2× the budget the model would likely cross the EP1 baseline-equivalent threshold, but I cannot prove it would beat 7.54% without running it.

2. **The slicing penalty is real.** `--model-slices 64` (forced by 96GB memory limits at 768d) approximately halved per-step throughput. Effective optimization steps were ~5442/epoch × 2.5 epochs ≈ 13,600, vs the SOTA baseline which gets ~19,500 steps in the same wall-clock with slices=128. Width cost ~30% of training steps.

3. **EP1=15.96% → EP2=10.03% is a 5.93pp jump in one epoch.** This matches the expected "bigger model takes longer to settle" pattern advisor described. EP3 partial telemetry showed continued negative slope but did not complete the full validation pass.

4. **Same hyperparameters may be poorly suited to 768d.** Lion lr=1e-4 was tuned for 12.7M params (4L/512d). Width scaling rules of thumb suggest Lion LR could be held roughly constant under µP, but the optimizer was tuned for the smaller model with no µP-style reparameterization in this codebase, so effective per-layer dynamics differ from baseline.

5. **Memory profile:** Peak GPU memory ≈ 84.7 GB / 96 GB on each rank. Comfortably fit at slices=64 — slices=128 OOM'd at step 0 (run `wp6g3kjz` aborted ~93.5 GB).

**Bottom line:** Within a 3-epoch budget, the cold-start 4L/768d/12h architecture is not competitive with the SOTA 4L/512d/8h baseline. The capacity gain doesn't pay back the training-time cost from gradient checkpointing within this budget. The τ_y/τ_z motivation was not borne out — those components got worse, not better.

### Exact command used

```bash
\
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --standalone --nproc_per_node=4 train.py \
  --agent norman \
  --wandb-group yi-round37-width-768d \
  --wandb-name "norman/4l-768d-12h-cold-start-slices64" \
  --epochs 3 \
  --learnable-pe --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --lr-warmup-epochs 1 \
  --grad-ema-alpha 0.5 \
  --no-compile-model \
  --model-layers 4 --model-hidden-dim 768 --model-heads 12 --model-slices 64 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --ema-decay 0.999
```

### Peak memory
~84.7 GB / 96 GB (88.3%) per rank, all 4 ranks similar.

### W&B run ID
`q03gty6i`  — earlier OOM was `wp6g3kjz` (slices=128, aborted at step 0).

### Suggested follow-ups

1. **Smaller width step (4L/640d/10h, ~19M params).** Likely fits without `--model-slices 64`, retaining slices=128 throughput. 1.5× capacity bump preserves the orthogonal-to-depth narrative without the checkpoint penalty. Probably the best near-term test of "does width help at all?".
2. **Resume-style budget for 768d.** If we want to fully test 768d, the only fair test is a 6-epoch budget (or warm-starting from a STRING-sep checkpoint). 3 epochs is not enough wall-clock for a 2.3× larger cold-start model.
3. **Lion LR sweep at 768d.** lr ∈ {1.0, 1.4, 2.0}×1e-4. Wider models can often take higher LR per Lion µP scaling, but only for 1-epoch screening to find a viable starting point before committing budget.
4. **Per-component loss weighting.** Independent of width, the τ_y/τ_z gaps consistently dominate the val_abupt metric (both at baseline and here). A direct attack — e.g. surface-shear loss weight 1.5–2× — may be more cost-effective than architectural scale-ups for closing those specific channels.
5. **Confirm 768d isn't broken.** The cold-start failure is consistent with under-training, not a wiring bug. The negative slopes and EP1→EP2 trajectory match expected behavior for a larger model. No code change needed before next attempt.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #659 because ADVISOR: Width-over-depth experiment (4L/768d/12h, run q03gty6i) closes here.

**Result summary**
- EP2 val_abupt = 10.0258%, test_abupt = 11.2020%
- Passes the 10.5% EP2 gate, but is +2.49 pp worse than yi SOTA (7.5373% val / 8.72% test)
- τ_y=13.30%, τ_z=14.35% — both worse than the baseline τ values (~9.87%, ~11.25%), so the width hypothesis is not validated

**Why the experiment under-delivered**
The OOM at slices=128 forced a fallback to slices=64, cutting training throughput by ~30%. Combined with a cold-start 3-epoch budget, the model (28M params vs. 10M baseline) simply had too few gradient updates to converge. The per-step loss slope at termination was still steeply negative (~−1.54%/1k steps), meaning the experiment was measuring an undertrained checkpoint, not a converged one.

**The width hypothesis is not falsified — it was not given a fair test.** The core idea (wider heads → better anisotropic τ representation) remains physically plausible but was budget-limited.

**Recommended follow-up (if width direction is revisited)**
- Use 4L/640d/10h instead of 768d/12h — fits within slices=128, recovers the ~30% throughput penalty
- Budget at least 10 epochs (not 3) to let the wider model converge
- Or: re-examine τ_y/τ_z loss weighting (currently live in frieren PR #669) as a lighter-weight approach to the same problem

Closing as a clean dead end — no cherry-pick needed, no bug fixes to preserve..
